core/src/Streamly/Internal/Data/Fold/Window.hs

```diff
@@ -319,34 +318,32 @@ maximum n = fmap (fmap snd) $ range n
 mean :: forall m a. (Monad m, Fractional a) => Fold m (a, Maybe a) a
 mean = Fold.teeWith (/) sum length
 
--- | Drop repeated elements within the specified tumbling window in the stream.
+-- | Drop repeated elements within the specified
+-- tumbling window in the stream.
 -- /Pre-release/
 --
 {-# INLINE nub #-}
-nub :: (Monad m, Ord a) => Fold m (a, Maybe a) (Set.Set a)
+nub :: (Monad m, Ord a) => Fold m (a, Maybe a) (Map.Map a Int)
 nub = Fold step initial extract
 
     where
 
     initial =
-        return $ Partial (Set.empty, Map.empty)
+        return $ Partial Map.empty
 
-    step (set, map0) (new, mOld) =
+    step map0 (new, mOld) =
         return $ Partial $
             case mOld of
                 Just a ->
                     let m1 = Map.insertWith (+) new (1::Int) map0
-                        (s1, m3) =
+                        m3 =
                             case Map.lookup a m1 of
                                 Just v ->
                                     if v == 1
-                                    then (Set.delete a set, m1)
-                                    else
-                                        let m2 = Map.adjust (\x -> x - 1) a m1
-                                            in (set, m2)
-                                _ -> (set, m1)
-                               in (Set.insert new s1, m3)
-                Nothing -> let m1 = Map.insertWith (+) new (1::Int) map0
-                               in (Set.insert new set, m1)
-
-    extract = return . fst
+                                    then Map.delete a m1
+                                    else Map.adjust (\x -> x - 1) a m1
```

Comment:
use a single `alterF` operation instead of using two operations (lookup and adjust).